File: example/kotlin/src/main/java/bzh/zomzog/kotlinworkshop/service/PonyService.java
package bzh.zomzog.kotlinworkshop.service;

import bzh.zomzog.kotlinworkshop.domain.Pony;
import bzh.zomzog.kotlinworkshop.domain.PonyBuilder;
import bzh.zomzog.kotlinworkshop.repository.PonyRepository;
import bzh.zomzog.kotlinworkshop.repository.documents.PonyMongo;
import org.bson.types.ObjectId;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class PonyService {

    private final PonyRepository ponyRepository;

    private final PonyMapper mapper;

    public PonyService(PonyRepository ponyRepository, PonyMapper mapper) {
        this.ponyRepository = ponyRepository;
        this.mapper = mapper;
    }

    public List<Pony> list() {
        return ponyRepository.findAll()
                .stream()
                .map(mapper::map)
                .collect(Collectors.toList());
    }

    public Pony save(Pony pony) {
        PonyMongo updated;
        if (null == pony.getId()) {
            updated = ponyRepository.save(mapper.map(pony));
        } else {
            final Optional<PonyMongo> fromDb = ponyRepository.findById(new ObjectId(pony.getId()));
            updated = fromDb.map(it -> mapper.update(it, pony))
                    .orElseThrow(() -> new RuntimeException("Pony not found")); // TODO Named exception
        }
        return mapper.map(updated);
    }

}

File: example/kotlin/src/main/kotlin/bzh/zomzog/kotlinworkshop/service/PonyService.kt
package bzh.zomzog.kotlinworkshop.service

import bzh.zomzog.kotlinworkshop.domain.Pony
import bzh.zomzog.kotlinworkshop.domain.PonyType
import bzh.zomzog.kotlinworkshop.repository.PonyRepository
import bzh.zomzog.kotlinworkshop.repository.documents.PonyMongo
import bzh.zomzog.kotlinworkshop.repository.documents.PonyMongoBuilder
import org.bson.types.ObjectId
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class PonyService(private val ponyRepository: PonyRepository,
                  private val mapper: PonyMapper) {

    fun list() = ponyRepository.findAll()
            .map { it.toPony() }

    fun save(pony: Pony): Pony {
        val updated = if (null == pony.id) {
//            ponyRepository.save(pony.toNewPonyMongo())
            mapper.mapNew(pony)
        } else {
            ponyRepository.findById(ObjectId(pony.id))
                    .map { mapper.update(it, pony) }
                    .orElseThrow { RuntimeException("Pony not found") } // TODO Named exception
        }
        val result = ponyRepository.save(updated)
        return mapper.map(result)
    }

    fun PonyMongo.toPony() = Pony(this.id.toString(), this.name, this.type, this.createdAt)
}